### ISSUE TITLE
Bug: Insufficiently declared dependency on `typing-extensions`

### ISSUE THREAD

**Issue by jmehnle** [FHPythonUtils/LicenseCheck#111]
- [X] I have checked for similar issues in the project's issue tracker.
- [X] I have searched closed issues to see if a similar problem was reported
  before.

### Description

`licensecheck/types.py` unconditionally imports from `typing_extensions`:

https://github.com/FHPythonUtils/LicenseCheck/blob/a89f9fc3cafbbdbedf03b3d2aa82ab3138544aaa/licensecheck/types.py#L8-L8

However, `pyproject.toml` does not at all declare a dependency on it, and `requirements.txt` requires it only on Pythons older than 3.11. As a result under Python 3.11 or later, `licensecheck` fails to execute with a traceback:

```
Traceback (most recent call last):
  File "/Users/julian/.cache/pre-commit/repoqpjhithw/py_env-python3.12/bin/licensecheck", line 5, in <module>
    from licensecheck import cli
  File "/Users/julian/.cache/pre-commit/repoqpjhithw/py_env-python3.12/lib/python3.12/site-packages/licensecheck/__init__.py", line 3, in <module>
    from licensecheck.cli import cli, main
  File "/Users/julian/.cache/pre-commit/repoqpjhithw/py_env-python3.12/lib/python3.12/site-packages/licensecheck/cli.py", line 15, in <module>
    from licensecheck import checker, fmt, license_matrix, packageinfo, types
  File "/Users/julian/.cache/pre-commit/repoqpjhithw/py_env-python3.12/lib/python3.12/site-packages/licensecheck/checker.py", line 11, in <module>
    from licensecheck import license_matrix
  File "/Users/julian/.cache/pre-commit/repoqpjhithw/py_env-python3.12/lib/python3.12/site-packages/licensecheck/license_matrix/__init__.py", line 46, in <module>
    from licensecheck.types import JOINS, ucstr
  File "/Users/julian/.cache/pre-commit/repoqpjhithw/py_env-python3.12/lib/python3.12/site-packages/licensecheck/types.py", line 8, in <module>
    from typing_extensions import Self
ModuleNotFoundError: No module named 'typing_extensions'
```

In my case, I actually had `typing-extensions` installed in my project's virtualenv, however I am trying to run `licensecheck` via a `pre-commit` hook, which creates its own virtualenv and installs only the dependencies stated by `licensecheck`, hence not installing `typing-extensions`.

## System Information

Please provide the following additional information about your system or
environment:

- Operating System (OS): macOS
- OS Version: 12.7.6
- Python 3.12 installed via MacPorts

## Proposed solution

Make the import from `typing_extensions` conditional on the Python version running, e.g.:

```python
import sys

if sys.version_info >= (3, 11):
    from typing import Self
else:
    from typing_extensions import Self
```

**Comment by FredHappyface** [comment]
Thanks so much for the issue 

No clue how this one slipped through! I've been testing on 3.11 on my Dev VM so yeah I'll get a fix out for this asap
